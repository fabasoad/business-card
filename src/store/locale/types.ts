export const LOAD_LOCALE = 'LOAD_LOCALE'
export const SET_LOCALE = 'SET_LOCALE'

export class Locale {
  code: string
  title: string

  constructor(code: string, title: string = null) {
    this.code = code
    this.title = ` ${title || code}`
  }
}

interface LoadLocaleAction {
  type: typeof LOAD_LOCALE
  payload: Locale
}

interface SetLocaleAction {
  type: typeof SET_LOCALE
  payload: Locale
}

export type LocaleActionTypes = LoadLocaleAction | SetLocaleAction

export type AppActions = LocaleActionTypes
