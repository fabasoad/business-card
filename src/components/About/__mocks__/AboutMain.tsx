import * as React from 'react'

export default function AboutMain() {
  return (
    <div data-testid="AboutMain"></div>
  )
}
