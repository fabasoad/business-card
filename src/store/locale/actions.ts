import i18n from 'i18next'
import { initReactI18next } from 'react-i18next'
import Backend from 'i18next-http-backend'
import LanguageDetector from 'i18next-browser-languagedetector'
import { Action } from 'redux'
import { AppState } from '../configureStore'
import { ThunkAction } from 'redux-thunk'
import { Dispatch } from 'react'
import { AppActions, Locale } from './types'
import SupportedLocales from '../../scripts/SupportedLocales'

type ThunkLocaleAction = ThunkAction<void, AppState, unknown, Action<string>>

const onLanguageChange = (t) => {
  document.title = t('business-card-title')
}

export const loadLocale = (payload: Locale): AppActions => ({
  type: 'LOAD_LOCALE',
  payload
})

export const startLoadLocale = (): ThunkLocaleAction =>
  async (dispatch: Dispatch<AppActions>) => {
    const params = new URLSearchParams(document.location.search)
    const code: string = params.get('locale') || localStorage.getItem('i18nextLng')
    const currentLocale = SupportedLocales.getOrDefault(code)
    if (document.location.search) {
      window.history.replaceState(
        {},
        document.title,
        document.location.origin + document.location.pathname + document.location.hash
      )
    }
    await i18n
      .use(Backend)
      .use(LanguageDetector)
      .use(initReactI18next)
      .init({
        backend: {
          loadPath: (process.env.NODE_ENV === 'development' ? '' : '/business-card') + '/locales/{{lng}}.json'
        },
        lng: currentLocale.code,
        fallbackLng: SupportedLocales.default().code,
        debug: process.env.NODE_ENV === 'development'
      }).then((t) => {
        onLanguageChange(t)
        dispatch(
          loadLocale({
            ...currentLocale
          })
        )
      })
  }

export const setLocale = (payload: Locale): AppActions => ({
  type: 'SET_LOCALE',
  payload
})

export const startSetLocale = (locale: Locale): ThunkLocaleAction =>
  (dispatch: Dispatch<AppActions>) => {
    i18n.changeLanguage(locale.code, (err, t) => {
      onLanguageChange(t)
      dispatch(
        setLocale({
          ...locale
        })
      )
    })
  }
