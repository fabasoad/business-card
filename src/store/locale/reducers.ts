import { LocaleActionTypes, SET_LOCALE, LOAD_LOCALE, Locale } from './types'
import SupportedLocales from '../../scripts/SupportedLocales'

export const localeReducer = (
  state = SupportedLocales.default,
  action: LocaleActionTypes
): Locale => {
  switch (action.type) {
    case LOAD_LOCALE:
      return {
        ...action.payload
      }
    case SET_LOCALE:
      return {
        ...action.payload
      }
    default:
      return state
  }
}
