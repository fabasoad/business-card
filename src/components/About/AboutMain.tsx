/// <reference path='./AboutMain.d.ts' />
import i18n from 'i18next'
import * as React from 'react'
import { connect } from 'react-redux'
import { ThunkDispatch } from 'redux-thunk'

import imgAboutMe from '../../assets/images/about_me.jpg'
import img1z0808Badge from '../../assets/images/badges/oracle-1z0-808.png'
import { Locale, AppActions } from '../../store/locale/types'
import { AppState } from '../../store/configureStore'
import ExperienceCalculations, {
  IExperienceCalculations
} from '../../scripts/ExperienceCalculations'

interface AboutMeProps {
  calc: IExperienceCalculations
}

type Props = AboutMeProps & LinkDispatchProps & LinkStateProps

export function AboutMain(props: Props) {
  const generalListItem1Keys = {
    totalExperience: props.calc.totalHumanize(props.locale.code),
    agileExperience: props.calc.agileHumanize(props.locale.code)
  }
  return (
    <div id="about">
      <div className="container">
        <div className="section-title text-center center">
          <h2>{i18n.t('business-card-about-me-title')}</h2>
          <hr />
        </div>
        <div className="row d-flex justify-content-center">
          <div className="about-photo col-md-12 d-flex justify-content-center">
            <img
              className="about-photo--portrait img-thumbnail"
              src={imgAboutMe}
              alt="About me"
            />
            <img
              className="about-photo--badge"
              src={img1z0808Badge}
              alt="Oracle 1Z0-808"
            />
          </div>
          <div className="col-md-8 col-md-offset-2 text-center">
            <div className="about-text">
              <ul className="list-unstyled mt-3">
                <li>
                  {i18n.t(
                    'business-card-about-me-general-list-item-1',
                    generalListItem1Keys
                  )}
                </li>
                <li>{i18n.t('business-card-about-me-general-list-item-2')}</li>
                <li>{i18n.t('business-card-about-me-general-list-item-3')}</li>
                <li>{i18n.t('business-card-about-me-general-list-item-4')}</li>
                <li>{i18n.t('business-card-about-me-general-list-item-5')}</li>
                <li>{i18n.t('business-card-about-me-general-list-item-6')}</li>
              </ul>
            </div>
          </div>
        </div>
      </div>
    </div>
  )
}

AboutMain.defaultProps = {
  calc: new ExperienceCalculations()
}

interface LinkStateProps {
  locale: Locale
}

interface LinkDispatchProps {
}

const mapStateToProps = (
  state: AppState,
  ownProps: AboutMeProps
): LinkStateProps => ({
  locale: state.locale
})

const mapDispatchToProps = (
  dispatch: ThunkDispatch<any, any, AppActions>,
  ownProps: AboutMeProps
): LinkDispatchProps => ({
})

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(AboutMain)
