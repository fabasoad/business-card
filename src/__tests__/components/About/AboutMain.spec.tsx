/// <reference types="jest" />
import * as React from 'react'
import { shallow } from 'enzyme'
import i18next from '../../__mocks__/i18next'
import { Locale } from '../../../store/locale/types'
import SupportedLocales from '../../../scripts/SupportedLocales'
import { AboutMain } from '../../../components/About/AboutMain'

test('should render AboutMain correctly', () => {
  const totalHumanize = (code: string) => `total-${code}`
  const agileHumanize = (code: string) => `agile-${code}`
  const totalHumanizeSpy = jest.fn(totalHumanize)
  const agileHumanizeSpy = jest.fn(agileHumanize)

  const locale: Locale = SupportedLocales.default
  const wrapper = shallow(<AboutMain locale={locale} calc={{
    agileHumanize: agileHumanizeSpy,
    totalHumanize: totalHumanizeSpy
  }} />)

  expect(wrapper).toMatchSnapshot()

  expect(totalHumanizeSpy).toHaveBeenCalledTimes(1)
  expect(totalHumanizeSpy).toHaveBeenCalledWith(locale.code)
  expect(agileHumanizeSpy).toHaveBeenCalledTimes(1)
  expect(agileHumanizeSpy).toHaveBeenCalledWith(locale.code)

  expect(i18next.t).toBeCalledTimes(7)
  expect(i18next.t).toHaveBeenCalledWith(
    expect.stringMatching(/^business-card-about-me-title$/)
  )
  expect(i18next.t).toHaveBeenCalledWith(
    expect.stringMatching(/^business-card-about-me-general-list-item-1$/),
    expect.objectContaining({
      totalExperience: totalHumanize(locale.code),
      agileExperience: agileHumanize(locale.code)
    })
  )
  for (let i = 2; i <= 6; i++) {
    expect(i18next.t).toHaveBeenCalledWith(
      expect.stringMatching(
        new RegExp(`^business-card-about-me-general-list-item-${i}$`)
      )
    )
  }
})
