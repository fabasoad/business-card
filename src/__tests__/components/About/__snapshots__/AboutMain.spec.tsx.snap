// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render AboutMain correctly 1`] = `
<div
  id="about"
>
  <div
    className="container"
  >
    <div
      className="section-title text-center center"
    >
      <h2>
        business-card-about-me-title
      </h2>
      <hr />
    </div>
    <div
      className="row d-flex justify-content-center"
    >
      <div
        className="about-photo col-md-12 d-flex justify-content-center"
      >
        <img
          alt="About me"
          className="about-photo--portrait img-thumbnail"
          src="test-file-stub"
        />
        <img
          alt="Oracle 1Z0-808"
          className="about-photo--badge"
          src="test-file-stub"
        />
      </div>
      <div
        className="col-md-8 col-md-offset-2 text-center"
      >
        <div
          className="about-text"
        >
          <ul
            className="list-unstyled mt-3"
          >
            <li>
              business-card-about-me-general-list-item-1
            </li>
            <li>
              business-card-about-me-general-list-item-2
            </li>
            <li>
              business-card-about-me-general-list-item-3
            </li>
            <li>
              business-card-about-me-general-list-item-4
            </li>
            <li>
              business-card-about-me-general-list-item-5
            </li>
            <li>
              business-card-about-me-general-list-item-6
            </li>
          </ul>
        </div>
      </div>
    </div>
  </div>
</div>
`;
